plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+        return objects.stream().filter(object -> !object.key().endsWith(PATH_SEPARATOR)).map(object -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + object.key(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public Stream<URI> listDirectories(URI dir)
+    {
+        List<CommonPrefix> prefixes = listS3Directory(dir).commonPrefixes();
+        return prefixes.stream().map(prefix -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + prefix.prefix(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public long size(URI uri, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        checkArgument(!isDirectory(uri), "expected a file URI but got a directory URI");
+        HeadObjectResponse response = getHeadObjectResponse(uri, secretKey);
+        if (response == null) {
+            throw new FileNotFoundException("File does not exist: " + uri);
+        }
+        return response.contentLength();
+    }
+
+    @Override
+    public int getWriteBufferSizeInBytes()
+    {
+        return multiUploadPartSize;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        try (Closer closer = Closer.create()) {
+            closer.register(deleteExecutor::shutdown);
+            closer.register(s3Client::close);
+            closer.register(s3AsyncClient::close);
+        }
+    }
+
+    private void deleteRecursivelySync(URI uri)
+    {
+        if (isDirectory(uri)) {
+            listDirectories(uri).forEach(this::deleteRecursively);
+            listFiles(uri).forEach(file -> deleteObject(getBucketName(file), keyFromUri(file)));
+            deleteObject(getBucketName(uri), keyFromUri(uri) + DIRECTORY_SUFFIX);
+        }
+        else {
+            deleteObject(getBucketName(uri), keyFromUri(uri));
+        }
+    }
+
+    private HeadObjectResponse getHeadObjectResponse(URI uri, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        HeadObjectRequest.Builder headObjectRequestBuilder = HeadObjectRequest.builder()
+                .bucket(getBucketName(uri))
+                .key(keyFromUri(uri));
+        S3RequestUtil.configureEncryption(secretKey, headObjectRequestBuilder);
+
+        try {
+            return s3Client.headObject(headObjectRequestBuilder.build());
+        }
+        catch (AwsServiceException e) {
+            if (e instanceof NoSuchKeyException) {
+                return null;
+            }
+            throw new IOException(e);
+        }
+    }
+
+    private ListObjectsV2Response listS3Directory(URI dir)
```

Comment:
Drop `S3` from name . As we are in `S3FileSystemExchangeStorage` it does not provide any extra information

Reply:
Also maybe rename to `listObjects` to match S3 API call.